AX: Merge AccessibilityList into AccessibilityRenderObject

hoffmanjoshua opened the pull request:
#### 80690759bef11e5cc189271860581864069c2b56
<pre>
AX: Merge AccessibilityList into AccessibilityRenderObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=297488">https://bugs.webkit.org/show_bug.cgi?id=297488</a>
<a href="https://rdar.apple.com/158440744">rdar://158440744</a>

Reviewed by Tyler Wilcock.

Merge the logic from AccessibilityList into AccessibilityNode- and
AccessibilityRenderObject, so that the AccessibilityList class can
be removed. This also adds a new helper, AXListHelpers, that does
list-specific work.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXListHelpers.cpp: Renamed from Source/WebCore/accessibility/AccessibilityList.cpp.
(WebCore::AXListHelpers::isAccessibilityList):
(WebCore::AXListHelpers::childHasPseudoVisibleListItemMarkers):
* Source/WebCore/accessibility/AXListHelpers.h: Renamed from Source/WebCore/accessibility/AccessibilityList.h.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::isAccessibilityList): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityList.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineListRoleWithCleanChildren):
(WebCore::shouldUseAccessibilityObjectInnerText):
(WebCore::AccessibilityNodeObject::isAccessibilityList const):
(WebCore::AccessibilityNodeObject::isUnorderedList const):
(WebCore::AccessibilityNodeObject::isOrderedList const):
(WebCore::AccessibilityNodeObject::isDescriptionList const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityList const):
(WebCore::AccessibilityObject::isListInstance const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::updateRoleAfterChildrenCreation):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: <a href="https://commits.webkit.org/298847@main">https://commits.webkit.org/298847@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ed045eb542ece92bed8afbe6511897502de3ee94

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116852 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36517 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27090 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122937 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/67442 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3286b865-6b5c-4cd8-b069-4319d3cad502) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118741 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/37214 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88743 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/43139 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/3ec28320-fceb-41f4-8b9c-508960ef6dae) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119801 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29668 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104824 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69201 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28732 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22929 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66593 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99047 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23083 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126064 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43752 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32865 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97409 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44115 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101024 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97207 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24752 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42527 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20475 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39768 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43637 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/49233 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43104 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/46443 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44809 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AXListHelpers.cpp:
```diff
@@ -43,68 +45,15 @@ namespace WebCore {
 {
     // Check if the list item has a pseudo-element that should be accessible (e.g. an image or text)
-    auto* element = dynamicDowncast<Element>(node);
+    RefPtr element = dynamicDowncast<Element>(&node);
```
Is this necessary? dynamicDowncast has an overload that takes a reference so it doesn't have to null-check

hoffmanjoshua replied:
Oops, fixed!

twilco commented on Source/WebCore/accessibility/AXListHelpers.cpp:
```diff
@@ -126,22 +75,13 @@ bool AccessibilityList::childHasPseudoVisibleListItemMarkers(const Node* node)
-    m_ariaRole = determineAriaRoleAttribute();
+    ASSERT(!listObject.needsToUpdateChildren() && listObject.childrenInitialized());
 
+    listObject.m_ariaRole = listObject.determineAriaRoleAttribute();
```
It breaks encapsulation a bit to set member variables for classes in helper functions. I feel like most of this method should be a member function on AccessibilityNodeObject — if we have to access fields or functions of an object, we shouldn't try to shoehorn that into a helper function like this. Maybe there are pieces that don't use member functions / fields that we can extract out into a helper?

hoffmanjoshua replied:
I ended up moving all of the logic from determineAccessibilityRoleWithCleanChildren into AccessibilityRenderObject, since in the loop itself we need to access private methods, so there's not much that can be extracted.

twilco replied:
Thanks! Yeah I think that makes sense. Any reason why we would move it to AccessibilityRenderObject instead of AccessibilityNodeObject? I think ultimately we want AXRenderObject to not exist, so anything we can put there now saves us more work later.

hoffmanjoshua replied:
We can put it in Node object, but for now I put it in the render variant since it's the only class that uses it. But I'll move it to Node

twilco commented on Source/WebCore/accessibility/AXObjectCache.h:
```diff
@@ -706,6 +706,8 @@ class AXObjectCache final : public CanMakeWeakPtr<AXObjectCache>, public CanMake
     WEBCORE_EXPORT static bool isAXThreadInitialized();
     WEBCORE_EXPORT RefPtr<AXIsolatedTree> getOrCreateIsolatedTree();
+
+    static bool isAccessibilityList(Element&);
```
Would this helper-style function be a good fit to put into the new AXListHelpers class, rather than in AXObjectCache?

hoffmanjoshua replied:
yes, moved it over there!

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -1435,7 +1435,7 @@ void AXObjectCache::handleChildrenChanged(AccessibilityObject& object)
 
     // The role of list objects is dependent on their children, so we'll need to re-compute it here.
-    if (is<AccessibilityList>(object))
+    if (object.isListInstance())
```
List "instance" isn't really accurate anymore — the word "instance" heavily implies an instance of subclass list, which doesn't exist anymore. Maybe just isList()? Or something similar?

hoffmanjoshua replied:
Renamed to `isAccessibilityList` per @minorninth 's comment below. AXCoreObject already has `isList()`, which is role-based.

minorninth commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2922,6 +2922,34 @@ bool AccessibilityNodeObject::canSetSelectedAttribute() const
     }
 }
 
+bool AccessibilityNodeObject::isListInstance() const
```
Should this just be isAccessibilityList since it delegates to a function with that name?

hoffmanjoshua replied:
Renamed!

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2922,6 +2922,34 @@ bool AccessibilityNodeObject::canSetSelectedAttribute() const
+    if (ariaRoleAttribute() == AccessibilityRole::Directory)
+        return true;
+
+    return elementName() == ElementName::HTML_ol;
```
Could this be written as:

`return ariaRoleAttribute() == AccessibilityRole::Directory || ariaRoleAttribute() == AccessibilityRole::Directory`

twilco commented on Source/WebCore/accessibility/AccessibilityRenderObject.h:
```diff
@@ -172,6 +172,8 @@ class AccessibilityRenderObject : public AccessibilityNodeObject {
     bool shouldIgnoreAttributeRole() const override;
     AccessibilityRole determineAccessibilityRole() override;
+    // Only used for lists.
+    AccessibilityRole determineAccessibilityRoleWithCleanChildren();
```
Feels like this comment would be redundant if the method name were more self-descriptive, e.g. `determineListRoleWithCleanChildren`

twilco reviewed (approved):
r+ with my last three comments addressed 🚀

hoffmanjoshua committed:
AX: Merge AccessibilityList into AccessibilityRenderObject
https://bugs.webkit.org/show_bug.cgi?id=297488
rdar://158440744

Reviewed by Tyler Wilcock.

Merge the logic from AccessibilityList into AccessibilityNode- and
AccessibilityRenderObject, so that the AccessibilityList class can
be removed. This also adds a new helper, AXListHelpers, that does
list-specific work.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXListHelpers.cpp: Renamed from Source/WebCore/accessibility/AccessibilityList.cpp.
(WebCore::AXListHelpers::isAccessibilityList):
(WebCore::AXListHelpers::childHasPseudoVisibleListItemMarkers):
* Source/WebCore/accessibility/AXListHelpers.h: Renamed from Source/WebCore/accessibility/AccessibilityList.h.
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
(WebCore::AXObjectCache::handleChildrenChanged):
(WebCore::isAccessibilityList): Deleted.
* Source/WebCore/accessibility/AXObjectCache.h:
* Source/WebCore/accessibility/AccessibilityList.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
(WebCore::AccessibilityNodeObject::determineListRoleWithCleanChildren):
(WebCore::shouldUseAccessibilityObjectInnerText):
(WebCore::AccessibilityNodeObject::isAccessibilityList const):
(WebCore::AccessibilityNodeObject::isUnorderedList const):
(WebCore::AccessibilityNodeObject::isOrderedList const):
(WebCore::AccessibilityNodeObject::isDescriptionList const):
* Source/WebCore/accessibility/AccessibilityNodeObject.h:
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityList const):
(WebCore::AccessibilityObject::isListInstance const): Deleted.
* Source/WebCore/accessibility/AccessibilityRenderObject.cpp:
(WebCore::AccessibilityRenderObject::computeIsIgnored const):
(WebCore::AccessibilityRenderObject::determineAccessibilityRole):
(WebCore::AccessibilityRenderObject::updateRoleAfterChildrenCreation):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: https://commits.webkit.org/298847@main
